File: examples_python/example_random.py
#!/usr/bin/env python3
"""
libzupt - Random Bytes and Hashing Example
Shows how to generate random bytes and compute hashes
"""

import sys
import os
sys.path.insert(0, os.path.dirname(__file__))

import zupt


def main():
    print("=" * 60)
    print("libzupt - Random Bytes and Hashing Example")
    print("=" * 60)
    print()

    # Generate random bytes
    print("1. Generating random bytes...")
    random_bytes = zupt.random_bytes(32)
    print(f"   Generated {len(random_bytes)} random bytes:")
    print(f"   {random_bytes.hex()}")
    print()

    # Generate a nonce
    print("2. Generating AES nonce...")
    nonce = zupt.random_bytes(zupt.AES_NONCE_SIZE)
    print(f"   Nonce ({len(nonce)} bytes): {nonce.hex()}")
    print()

    # Compute SHA-256 hash
    print("3. Computing SHA-256 hash...")
    data = b"Hello, Post-Quantum World!"
    sha256_hash = zupt.sha256(data)
    print(f"   Data: {data}")
    print(f"   SHA-256: {sha256_hash.hex()}")
    print()

    # Compute SHA3-512 hash
    print("4. Computing SHA3-512 hash...")
    sha3_512_hash = zupt.sha3_512(data)
    print(f"   Data: {data}")
    print(f"   SHA3-512: {sha3_512_hash.hex()}")
    print()

    # Use random bytes for key generation simulation
    print("5. Simulating key derivation...")
    salt = zupt.random_bytes(16)
    print(f"   Salt: {salt.hex()}")
    derived_key = zupt.sha256(salt + b"my-secret-password")
    print(f"   Derived key (32 bytes): {derived_key.hex()}")
    print()

    print("=" * 60)
    print("Random bytes and hashing example passed!")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: examples_java/src/com/libzupt/ExampleRandom.java
package com.libzupt;

import java.util.Formatter;

/**
 * Random Bytes and Hashing Example.
 * Shows random number generation and hashing functions.
 */
public class ExampleRandom {

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("libzupt - Random Bytes and Hashing Example");
        System.out.println("=".repeat(60));
        System.out.println();

        // Generate random bytes
        System.out.println("1. Generating random bytes...");
        byte[] randomBytes = NativeLib.randomBytes(32);
        System.out.println("   Generated 32 random bytes:");
        System.out.println("   " + bytesToHex(randomBytes));
        System.out.println();

        // Generate AES nonce
        System.out.println("2. Generating AES nonce...");
        byte[] nonce = NativeLib.randomBytes(16);
        System.out.println("   Nonce (16 bytes): " + bytesToHex(nonce));
        System.out.println();

        // Compute SHA-256 hash
        System.out.println("3. Computing SHA-256 hash...");
        byte[] data = "Hello, Post-Quantum World!".getBytes();
        byte[] sha256Hash = NativeLib.sha256(data);
        System.out.println("   Data: " + new String(data));
        System.out.println("   SHA-256: " + bytesToHex(sha256Hash));
        System.out.println();

        // Compute SHA3-512 hash
        System.out.println("4. Computing SHA3-512 hash...");
        byte[] sha3_512Hash = NativeLib.sha3512(data);
        System.out.println("   Data: " + new String(data));
        System.out.println("   SHA3-512: " + bytesToHex(sha3_512Hash));
        System.out.println();

        // Simulate key derivation
        System.out.println("5. Simulating key derivation...");
        byte[] salt = NativeLib.randomBytes(16);
        byte[] password = "my-password".getBytes();
        // Simple KDF simulation: SHA256(salt || password)
        byte[] combined = new byte[salt.length + password.length];
        System.arraycopy(salt, 0, combined, 0, salt.length);
        System.arraycopy(password, 0, combined, salt.length, password.length);
        byte[] derivedKey = NativeLib.sha256(combined);

        System.out.println("   Salt: " + bytesToHex(salt));
        System.out.println("   Derived key (32 bytes): " + bytesToHex(derivedKey));
        System.out.println();

        System.out.println("=".repeat(60));
        System.out.println("Random bytes and hashing example passed!");
        System.out.println("=".repeat(60));
    }

    private static String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }
}